src/main/java/seedu/address/model/tag/MedicationMap.java

```diff
@@ -0,0 +1,98 @@
+package seedu.address.model.tag;
+
+import static seedu.address.logic.commands.CountCommand.EACH_MEDICATION_COUNT;
+import static seedu.address.logic.commands.CountCommand.MEDICATION_COUNT;
+
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableMap;
+import seedu.address.model.person.Person;
+
+/**
+ * Encapsulates a map of medications and the number of times they have been assigned to patients.
+ */
+public class MedicationMap {
+    public static final String COUNT_BY_MEDICATION = "Patient count by medication:";
+    private final ObservableMap<String, Integer> medicationMap = FXCollections.observableHashMap();
+
+    /**
+     * Adds a medication to the medication map.
+     * @param medication The medication to be added.
+     */
+    public void add(Medication medication) {
+        medicationMap.putIfAbsent(medication.medicationName, 0);
+        medicationMap.put(medication.medicationName, medicationMap.get(medication.medicationName) + 1);
+    }
+
+    /**
+     * Removes a medication from the medication map.
+     * @param medication The medication to be removed.
+     */
+    public void remove(Medication medication) {
+        if (medicationMap.containsKey(medication.medicationName)) {
+            medicationMap.put(medication.medicationName, medicationMap.get(medication.medicationName) - 1);
+        }
+        medicationMap.entrySet().removeIf(entry -> entry.getValue() <= 0);
+    }
+
+    /**
+     * Adds a person's medications to the medication map.
+     * @param person The person whose medications are to be added.
+     */
+    public void addMedicationsOfPatient(Person person) {
+        person.getMedications().forEach(this::add);
+    }
+
+    /**
+     * Clears the medication map of medications allocated to the patient.
+     * @param patient The patient to be removed.
+     */
+    public void clearMedicationsOfPatient(Person patient) {
+        patient.getMedications().forEach(this::remove);
+    }
```

Comment:
Just a slight change, I believe it's better to use the same parameter name for both `MedicationMap::addMedicationsOfPatient` and `MedicationMap::clearMedicationsOfPatient`.

Either use `Person person` or `Person patient` for both.